### PR TITLE
Generalize fixplate.

### PR DESCRIPTION
**Pull request by sellout**
* Rename to `recursionschemes`,
* move folds and unfolds into type classes,
* merge `FoldableT` into the `Recursive` type class, and
* rename `Term` to `Fix`.

### CODE REVIEW

**Comment by sellout** [comment]
It may help to look at sellout/slamengine@e4e13cb0be0571529265ff990e3785803ff85257, because the other commit just renames `Term` and adds a ton of noise to the diff.

**Comment by jonsterling** [comment]
Strong agreement on renaming `Term` to `Fix` :smiley:

**Comment by mossprescott** [comment]
Mostly looks great, and if there's nothing to be done about the new type annotations, then I can live with them.

:+1:

File: core/src/main/scala/slamdata/recursionschemes/package.scala [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+  }
+  implicit val FixCorecursive = new Corecursive[Fix] {
+    def embed[F[_]](t: F[Fix[F]]) = Fix(t)
+  }
+
+  implicit def CofreeRecursive[A] = new Recursive[Cofree[?[_], A]] {
+    def project[F[_]](t: Cofree[F, A]) = t.tail
+  }
+
+  // NB: Not currently possible because this requires `Functor[F]` and that
```
Comment (wemrysi) [inline core/src/main/scala/slamdata/recursionschemes/package.scala]: Should be possible once we're able to upgrade to scalaz-7.2 due to `Free` refactoring.

File: core/src/main/scala/slamdata/recursionschemes/package.scala [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+      val t = RF.render(v.unFix)
+        NonTerminal(t.nodeType, t.label, v.children.map(render(_)))
+    }
+  }
+
+  implicit def FixEqual[F[_]](implicit F: Equal ~> λ[α => Equal[F[α]]]):
+    Equal[Fix[F]] =
+  Equal.equal { (a, b) => F(FixEqual[F]).equal(a.unFix, b.unFix) }
+
+  sealed trait Hole
```
Comment (wemrysi) [inline core/src/main/scala/slamdata/recursionschemes/package.scala]: What is the purpose of this type? It doesn't seem to be used anywhere.

Comment (sellout) [inline core/src/main/scala/slamdata/recursionschemes/package.scala]: I’m not sure. The functions below it come from http://hackage.haskell.org/package/fixplate-0.1.5/docs/Data-Generics-Fixplate-Open.html#g:2 but I don’t see any reference to anything like `Hole` in that code. @jdegoes – any recollection?

Comment (wemrysi) [inline core/src/main/scala/slamdata/recursionschemes/package.scala]: Looks like it is referenced in the `Shape` and the aforementioned link says it is useful for drawing trees as well.

**Comment by wemrysi** [comment]
:+1: Nice work!

**Comment by sellout** [comment]
@mossprescott – I think we can do something about the new type annotations eventually.

With something like Scalaz’s `…IdV` traits to inject the unfolds into `A`,
```scala
Corecursive[Term].futu(81)(extract2and3)
```
hopefully becomes
```scala
81.futu[Term](extract2and3)
```

And @tpolecat’s [partially-applied type constructors](http://tpolecat.github.io/2015/07/30/infer.html) can help reduce the not-new annotations that need to be provided on things like `cataM`.

**Pull request merged**